### PR TITLE
[Bug 16203] Parsing a Uint list fails in LC 7 if it contains "empty"

### PR DESCRIPTION
**Pull request by livecodepanos**

### CODE REVIEW

**Comment by runrevmark** [comment]
We need to double check whether this is appropriate. If the hilitedLines property used its own code to parse the list pre-7 (it probably did) then we need a separate property type which indicates a list of UInts where empty items are ignored.

Can you survey the properties which use ListOfUInt, then look back at 6.7 and see which ones accept skipping empty elements, and ones which don't?

**Comment by livecodepanos** [comment]
Ok, I have found 4 properties that use ItemsOfUint. These are the `hilitedLines`, `tabStops`, `tabWidths`, `dashes`. Each one has a different behaviour regarding the empty elements:

`set the hilitedLines of fld 1 to empty,4,empty,empty,5`
`put the hilitedLines of fld 1` --> returns 4,5

`set the tabStops of fld 1 to empty,50,empty,empty,80,120`
`put the tabStops of fld 1` --> returns 0,50,50,50,80,120

`set the tabWidths of fld 1 to empty,50,empty,empty,80,120`
`put the tabWidths of fld 1` --> returns 0,50,0,0,80,120

`set the dashes of grc 1 to empty,50,empty,empty,80,120`
`put the dashes of grc 1` --> returns 0,50,0,0,80,120

So:
- the `hilitedLines` skip empty elements, 
- the `tabStops` map empty to 0 when empty is in the beginning and to the previous element when empty is not in the beginning, 
- the `tabWidths` and `dashes` map empty to 0 always

**Comment by runrevmark** [comment]
Okay - so the odd one out here appears to be 'the hilitedLines'... However, we should be able to handle the exception in the field setter for 'hilitedLines'. I _think_ 0's are ignored (or were previously) - so as long as the field setter ignores any 0's in the list, we can leave ParseUIntList as it is.

**Comment by livecodepanos** [comment]
The problem is that if the list contains "empty", either in the beginning or not, then the list is parsed up to point where "empty" is met. For example if doing `set the hilitedlines of fld 1 to empty,2`, then `MCField::SetHilitedLines` with be called with `p_count == 0` and `p_lines == NULL`. I think the best solution would be to add a separate property type which indicates a list of UInts where empty items are ignored, i.e. as you suggested in your first comment. I'll update this PR shortly.

**Comment by runrevmark** [comment]
@livecodepanos: I don't believe that is what I found when I tried this out...

If I did 'set the hilitedLines of field 1 to "1,,3" then I get lines 1 and 3 hilited (in 6.7.x).

So - I think the 'UIntList' is correct... It should parse a list of integers - if empty is encountered it returns '0' in that position. This means the other properties relying on UIntList will still work. Then, the tweak that is needed is that in the hilitedLines setter we need to make sure it ignores lines of index 0.

**Comment by livecodepanos** [comment]
Sorry for the misunderstanding. In 6.7.x it works as expected. The problem is specific to LC 7.x.

**Comment by runrevmark** [comment]
@livecodepanos: Right so the bug here is that ItemsOfUInt shouldn't be stopping at an empty item. If it is only used by those four properties, then it should return 0 for empty.

**Comment by livecodepanos** [comment]
Ok, I did some changes. I renamed Lines/Items/MixedItemsOfUInt to LooseUInt, to indicate that empty is interpreted as 0 (same for Double), and amended MCPropertyParse*List to treat an empty item as 0. I also modified the setter of 'enabledtracks' to skip empty (--> 0) items, to match the behaviour of LC 6.7.x.

File: engine/src/exec.cpp [hunk 1]
```diff
@@ -1573,14 +1573,32 @@ static bool MCPropertyParseUIntList(MCStringRef p_input, char_t p_delimiter, uin
 	uindex_t t_new_offset;
 	t_new_offset = 0;
 	
+	// PM-2015-10-13: [[ Bug 16203 ]] Replace any "empty" elements with 0 in the beginning of the list
+	if (MCStringGetNativeCharAtIndex(p_input, 0) == p_delimiter)
```
Comment (runrevmark) [inline engine/src/exec.cpp]: I'm not sure this clause is needed if you remove the 't_new_offset <= t_old_offset' check from the while condition

File: engine/src/exec.cpp [hunk 2]
```diff
@@ -1624,11 +1642,28 @@ static bool MCPropertyParseDoubleList(MCStringRef p_input, char_t p_delimiter, u
...
+	// PM-2015-10-13: [[ Bug 16203 ]] Replace any "empty" elements with 0.0 in the beginning of the list
+	if (MCStringGetNativeCharAtIndex(p_input, 0) == p_delimiter)
+	{
+		t_old_offset++;
+		t_new_offset++;
+		
+		t_success = t_list . Push(0.0);
+	}
+	
 	while (t_success && t_old_offset <= t_length)
```
Comment (runrevmark) [inline engine/src/exec.cpp:1634]: Looking at this loop - I think we can make this work in a slightly simpler way. The 'bug' in the current code is the t_new_offset <= t_old_offset breaking out after FirstIndexOfChar.

I think if we make the logic:

```
while(t_old_offset <= t_length)
{
  if (!firstindexofchar(delimiter, t_old_offset -> t_new_offset))
  {
    if (t_old_offset == t_length)
      break;
    t_new_offset = t_length;
  }
  if (t_new_offset == t_old_offset)
    Push(0)
  else
    Push(attempt to convert item as number)

  t_old_offset = t_new_offset + 1;
}
```

Then it should work...

Basically we want to treat any empty item as 0 - unless it is the trailing item (i.e. "0,,1," - should be a 3 element result - 0,0,1 - as the rule is that trailing delimiters are ignored).

**Comment by runrevmark** [comment]
The names of the MCPropertyParse method should also be updated to be ParseLoose

**Comment by runrevmark** [comment]
Great work @livecodepanos.

@livecode-vulcan review ok 963b311

**Comment by runrevmark** [comment]
@livecode-vulcan review ok c8d3bd7

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok c8d3bd793b7361ceb8aaf499156960c0950b30df

**Comment by livecode-vulcan** [comment]
😞 test failure c8d3bd793b7361ceb8aaf499156960c0950b30df
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/870)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/884)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/871)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/874)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/884)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/894)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/887)

**Comment by livecodepanos** [comment]
@livecode-vulcan retry c8d3bd7

**Comment by livecode-vulcan** [comment]
😎 test success c8d3bd793b7361ceb8aaf499156960c0950b30df
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/900)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/909)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/898)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/901)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/910)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/920)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/914)

**Pull request merged**